flutter-studio/src/io/flutter/android/AndroidModuleLibraryManager.java

```diff
@@ -0,0 +1,198 @@
+ * into the project, so full editing support is available.
+ *
+ * @see AndroidModuleLibraryType
+ * @see AndroidModuleLibraryProperties
+ */
+public class AndroidModuleLibraryManager extends AbstractLibraryManager<AndroidModuleLibraryProperties> {
+  private static final Logger LOG = Logger.getInstance(AndroidModuleLibraryManager.class);
+  private static final String BUILD_FILE_NAME = "build.gradle";
+  private final AtomicBoolean isUpdating = new AtomicBoolean(false);
+
+  public AndroidModuleLibraryManager(@NotNull Project project) {
+    super(project);
+  }
+
+  public void update() {
+    doGradleSync(getProject(), (Project x) -> updateAndroidLibraryContent(x));
+  }
+
+  private Void updateAndroidLibraryContent(@NotNull Project androidProject) {
+    LibraryTable androidProjectLibraryTable = LibraryTablesRegistrar.getInstance().getLibraryTable(androidProject);
+    Library[] androidProjectLibraries = androidProjectLibraryTable.getLibraries();
+    if (androidProjectLibraries.length == 0) {
+      LOG.warn("Gradle sync was incomplete -- no Android libraries found");
+      return null;
+    }
+    HashSet<String> urls = new HashSet<>();
+    for (Library refLibrary : androidProjectLibraries) {
+      urls.addAll(Arrays.asList(refLibrary.getRootProvider().getUrls(OrderRootType.CLASSES)));
+    }
+    updateLibraryContent(urls);
+    return null;
+  }
+
+  @NotNull
+  @Override
+  protected String getLibraryName() {
+    return LIBRARY_NAME;
+  }
+
+  @NotNull
+  @Override
+  protected PersistentLibraryKind<AndroidModuleLibraryProperties> getLibraryKind() {
+    return LIBRARY_KIND;
+  }
+
+  private void scheduleUpdate() {
+    if (isUpdating.get()) {
+      return;
+    }
+
+    final Runnable runnable = this::updateAndroidLibraries;
+    DumbService.getInstance(getProject()).smartInvokeLater(runnable, ModalityState.NON_MODAL);
+  }
+
+  private void updateAndroidLibraries() {
+    if (!isUpdating.compareAndSet(false, true)) {
+      return;
+    }
+    update();
+  }
+
+  private void doGradleSync(Project flutterProject, Function<Project, Void> callback) {
+    // TODO(messick): Collect URLs for all Android modules, including those within plugins.
+    VirtualFile dir = flutterProject.getBaseDir().findChild("android");
+    assert (dir != null);
+    EmbeddedAndroidProject androidProject = new EmbeddedAndroidProject(FileUtilRt.toSystemIndependentName(dir.getPath()), null);
+    androidProject.init();
+
+    GradleSyncListener listener = new GradleSyncListener() {
+      @SuppressWarnings("override")
+      public void syncTaskCreated(@NotNull Project project, @NotNull GradleSyncInvoker.Request request) {}
+
+      @Override
+      public void syncStarted(@NotNull Project project, boolean skipped, boolean sourceGenerationRequested) {}
+
+      @Override
+      public void setupStarted(@NotNull Project project) {}
+
+      @Override
+      public void syncSucceeded(@NotNull Project project) {
+        callback.apply(androidProject);
```

Comment:
Does it make sense to put the flag update in a finally?  Something like:

```java
try {
  callback.apply(androidProject);
} finally {
  isUpdating.set(false);
}
```

EDIT: I'm probably over-thinking.

Reply:
No, that doesn't work too well because the callback runs async. Effectively removes the flag. Speaking from experience here :)